### PR TITLE
Isito agent sets the grace period based on cert TTL instead of client-side settings

### PR DESCRIPTION
**Pull request by myidpt**
Addresses https://github.com/istio/istio/issues/22226
Also, enable istiod to apply a default workload cert TTL when the client sends a request with non-positive lifetime.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pkg/istio-agent/sds-agent.go [hunk 1]
```diff
@@ -448,7 +448,7 @@ func applyEnvVars() {
 	serverOptions.Pkcs8Keys = pkcs8KeysEnv
 	serverOptions.RecycleInterval = staledConnectionRecycleIntervalEnv
 	workloadSdsCacheOptions.SecretTTL = secretTTLEnv
-	workloadSdsCacheOptions.SecretRefreshGraceDuration = secretRefreshGraceDurationEnv
+	workloadSdsCacheOptions.SecretRefreshGracePeriodRatio = secretRefreshGracePeriodRatioEnv
```
Comment (lei-tang) [inline pkg/istio-agent/sds-agent.go]: Where is SecretRefreshGracePeriodRatio used? I did a search in the PR and only found the places it is defined/declared, but not where it is used.

Comment (myidpt) [reply pkg/istio-agent/sds-agent.go]: It turned out to be a typo. Should be SecretRotationGracePeriodRatio.

File: security/pkg/nodeagent/cache/secretcache.go [hunk 2]
```diff
@@ -807,9 +808,11 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token string, connKey
...
 }
 
-func (sc *SecretCache) shouldRefresh(s *model.SecretItem) bool {
-	// secret should be refreshed before it expired, SecretRefreshGraceDuration is the grace period;
-	return time.Now().After(s.ExpireTime.Add(-sc.configOptions.SecretRefreshGraceDuration))
+func (sc *SecretCache) shouldRotate(secret *model.SecretItem) bool {
+	// secret should be rotated before it expired.
+	secretLifeTime := secret.ExpireTime.Sub(secret.CreatedTime)
+	gracePeriod := secretLifeTime * time.Duration(sc.configOptions.SecretRotationGracePeriodRatio)
+	return time.Now().After(secret.ExpireTime.Add(-gracePeriod))
```
Comment (lei-tang) [inline security/pkg/nodeagent/cache/secretcache.go]: Is "-gracePeriod" equivalent to "-1*gracePeriod"? Is secret.ExpireTime.Add(-gracePeriod) equivalent to secret.ExpireTime.Sub(gracePeriod)?

Comment (myidpt) [reply security/pkg/nodeagent/cache/secretcache.go]: Nope. ExpireTime.Sub(gracePeriod) won't work. It doesn't take a Duration as a parameter but Add() does.

File: security/pkg/nodeagent/cache/secretcache.go [hunk 3]
```diff
@@ -501,30 +501,30 @@ func (sc *SecretCache) rotate(updateRootFlag bool) {
...
 				return true
 			}
 
 			atomic.AddUint64(&sc.rootCertChangedCount, 1)
-			t := time.Now()
+			now := time.Now()
 			ns := &model.SecretItem{
 				ResourceName: connKey.ResourceName,
 				RootCert:     sc.rootCert,
 				ExpireTime:   sc.rootCertExpireTime,
```
Comment (JimmyCYJ) [inline security/pkg/nodeagent/cache/secretcache.go]: This is not related to the PR, but do we guarantee that root cert in sc.rootCert is already rotated at this point? Are there any tests covering this part?

Comment (myidpt) [reply security/pkg/nodeagent/cache/secretcache.go]: Good question. Can you help to check? If you find any problem, please file an issue.

**Pull request merged**